tests/saitests/sai_qos_tests.py

```diff
@@ -2023,6 +2023,97 @@ def runTest(self):
         finally:
             sai_thrift_port_tx_enable(self.client, asic_type, [dst_port_id])
 
+class PGDropTest(sai_base_test.ThriftInterfaceDataPlane):
+    def runTest(self):
+        time.sleep(5)
+        switch_init(self.client)
+
+        # Parse input parameters
+        dscp = int(self.test_params['dscp'])
+        ecn = int(self.test_params['ecn'])
+        router_mac = self.test_params['router_mac']
+        sonic_version = self.test_params['sonic_version']
+        pg = int(self.test_params['pg'])
+        queue = int(self.test_params['queue'])
+        dst_port_id = int(self.test_params['dst_port_id'])
+        dst_port_ip = self.test_params['dst_port_ip']
+        dst_port_mac = self.dataplane.get_mac(0, dst_port_id)
+        src_port_id = int(self.test_params['src_port_id'])
+        src_port_ip = self.test_params['src_port_ip']
+        src_port_vlan = self.test_params['src_port_vlan']
+        src_port_mac = self.dataplane.get_mac(0, src_port_id)
+        asic_type = self.test_params['sonic_asic_type']
+        pkts_num_trig_pfc = int(self.test_params['pkts_num_trig_pfc'])
+        # Should be set to cause at least 1 drop at ingress
+        pkts_num_trig_ingr_drp = int(self.test_params['pkts_num_trig_ingr_drp'])
+        iterations = int(self.test_params['iterations'])
+        margin = int(self.test_params['pkts_num_margin'])
+
+        hwsku = self.test_params['hwsku']
+        pkt_dst_mac = router_mac if router_mac != '' else dst_port_mac
+
+        # get counter names to query
+        ingress_counters, egress_counters = get_counter_names(sonic_version)
+
+        # Prepare IP packet data
+        ttl = 64
+        packet_length = 64
+        pkt = construct_ip_pkt(packet_length,
+                               pkt_dst_mac,
+                               src_port_mac,
+                               src_port_ip,
+                               dst_port_ip,
+                               dscp,
+                               src_port_vlan,
+                               ecn=ecn,
+                               ttl=ttl)
+
+        print >> sys.stderr, "test dst_port_id: {}, src_port_id: {}, src_vlan: {}".format(
+            dst_port_id, src_port_id, src_port_vlan
+        )
+
+        try:
+            pass_iterations = 0
```

Comment:
It would be better to check the fail_iterations and add an assert at the end of each iteration. In this way, this case could exist at the earliest time, and don't need to waste time iterating all 100 times.

Reply:
The test runtime has been optimized, should run in about 100 seconds. 

The test also is designed to only assert at the end of the iterations after the summary statistic has been produced.